tests/test_publish_table.py

```diff
@@ -0,0 +1,361 @@
+import os
+
+import numpy as np
+import pyarrow.parquet as parquet
+import pytest
+
+import audeer
+import audformat
+import audiofile
+
+import audb
+
+
+@pytest.fixture
+def db(tmpdir) -> str:
+    r"""Database fixture.
+
+    Creates a minimal database,
+    containing a media file
+    and a table.
+    The database is not yet stored to disk.
+
+    Args:
+        tmpdir: tmpdir fixture
+
+    Returns:
+        database object, and build dir path
+
+    """
+    name = "test-db"
+    file = "data/file1.wav"
+    table = "files"
+    build_dir = audeer.path(tmpdir, "build")
+    data_dir = audeer.mkdir(build_dir, os.path.dirname(os.path.normpath(file)))
+    audio_file = audeer.path(data_dir, os.path.basename(os.path.normpath(file)))
+    signal = np.zeros((2, 1000))
+    sampling_rate = 8000
+    audiofile.write(audio_file, signal, sampling_rate)
+    db = audformat.Database(name)
+    db.schemes["speaker"] = audformat.Scheme("str")
+    index = audformat.filewise_index([file])
+    db[table] = audformat.Table(index)
+    db[table]["speaker"] = audformat.Column(scheme_id="speaker")
+    db[table]["speaker"].set(["adam"])
+    db.meta["build_dir"] = build_dir
+
+    yield db
+
+
+def expected_table_checksum(path: str) -> str:
+    r"""Expected checksum of table file.
+
+    MD5 sum for CSV table,
+    ``"hash"`` metadata entry for PARQUET files.
+
+    Args:
+        path: path to table file
+
+    """
+    if audeer.file_extension(path) == "parquet":
+        # See https://github.com/audeering/audformat/pull/419
+        return parquet.read_schema(path).metadata[b"hash"].decode()
+    else:
+        return audeer.md5(path)
+
+
+@pytest.mark.parametrize("storage_format", ["csv", "parquet"])
+def test_publish_table_storage_format(db, repository, storage_format):
```

Comment:
I wonder whether this tests could be made more modular and therefore readable. For example, `test_publish_table_storage_format` has a few sequences of assertions + modifications

Grepping out the comments I get

```

214     # Save table as PARQUET and CSV at the same time
222     # ASSERT: Check database build_dir looks as expected
227     # Publish database
231     # ASSERT: Check files are published to repository
245     # Update only CSV table
254     # Remove PKL file to ensure CSV file is not newer
257     # ASSERT Check database build_dir looks as expected
263     # Publishing updated database
267     # Check files are published to repository.
268     # Only header and dependency table change, + UNCOMMNENTED ASSERT
269     # as the PARQUET table stayed the same
280     # Load database to cache + UNCOMMENTED ASSERT

```

Would it maybe makse sense to turn it into a test class with modifications turned into class-local fixtures? Probably having a one-to-one equivalence of test case and assertions would go too far, but with fixtures for a specific test class it would be easier to read as it would be clear at what stage the database is in when an  assertion is made.

Reply:
It is indeed nicer that the tests are better separated then, but the downside is that a lot of code needs to repeated (or I'm not very good at finding a nice class based solution ;) ).

This is what I came up with so far as a replacement:

```python
@pytest.mark.parametrize("storage_format", ["csv", "parquet"])
def TestPublishTableStorageFormat():
    r"""Test publishing and of tables for different storage formats.

    Tables are stored as CSV or PARQUET files.
    ``audb`` is handling them slightly different:

    * CSV tables are published as ``<table-id>.zip``,
      and the table checksum is calculated
      on the original CSV file
    * PARQUET tables are published as ``<table-id>.parquet``
      and the table checksum is read
      from the header of the PARQUET file

    Args:
        db: db fixture
        build_dir: build_dir fixture
        repository: repository fixture,
            providing a non-persistent repository
            on a file-system backend
        storage_format: storage format of tables

    """

    def other_table_file(self, db: audformat.Database, storage_format: str) -> str:
        r"""Filename of first database table with other storage format.

        csv => parquet
        parquet => csv

        Args:
            db: database
            storage_format: selected storage format

        Returns:
            table filename with other storage format

        """
        table = list(db)[0]
        if storage_format == "csv":
            ext = "parquet"
        elif storage_format == "parquet":
            ext = "csv"
        return f"db.{table}.{ext}"

    def media_file(self, db: audformat.Database) -> str:
        r"""Filename of first media file in database.

        Args:
            db: database

        Returns:
            filename of media file

        """
        return db.files[0]

    def publish(self, version: str, storage_format: str) -> audb.Dependencies:
        r"""Publish database.

        Args:
            version: version of database
            storage_format: storage format of database tables

        Returns:
            dependency object of published database

        """
        table = list(db)[0]
        file = db.files[0]
        audio_file = audeer.path(build_dir, file)
        db.save(build_dir, storage_format=storage_format)
        deps = audb.publish(build_dir, version, repository)
        return deps

    def table_file(self, db: audformat.Database, storage_format: str) -> str:
        r"""Table file name of first table in database.

        Args:
            db: database
            storage_format: storage format of databas tables

        Returns:
            table file name

        """
        table = list(db)[0]
        return f"db.{table}.{storage_format}"

    def test_database_save(self, build_dir, db, storage_format):
        r"""Test correct files are stored to build dir.

        Args:
            build_dir: build dir fixture
            db: database fixture
            storage_format: storage format of database tables

        """
        db.save(build_dir, storage_format=storage_format)
        assert os.path.exists(
            audeer.path(build_dir, self.table_file(db, storage_format))
        )
        assert not os.path.exists(
            audeer.path(build_dir, self.other_table_file(db, storage_format))
        )
        assert os.path.exists(self.media_file(db))

    def test_database_publish(
        self,
        build_dir: str,
        db: audformat.Database,
        repository: audb.Repository,
        storage_format: str,
    ):
        r"""Tests files stored in repository and dependency table entries.

        Tables are stored differently in the repository
        for different storage formats,
        and their dependency table entry looks different as well.

        Args:
            build_dir: build dir fixture
            db: database fixture
            repository: repository fixture
            storage_format: database table storage format

        """
        # Publish database
        db.save(build_dir, storage_format=storage_format)
        version = "1.0.0"
        deps = audb.publish(build_dir, version, repository)

        # Check files are published to repository
        repo = audeer.path(repository.host, repository.name)
        file = self.media_file(db)
        table = list(db)[0]
        dependency_file = "db.parquet"
        header_file = "db.yaml"
        media_file = f"{deps.archive(file)}.zip"
        if storage_format == "csv":
            meta_file = f"{table}.zip"
        elif storage_format == "parquet":
            meta_file = f"{table}.parquet"
        expected_paths = [
            audeer.path(repo, db.name, version, dependency_file),
            audeer.path(repo, db.name, version, header_file),
            audeer.path(repo, db.name, "media", version, media_file),
            audeer.path(repo, db.name, "meta", version, meta_file),
        ]
        assert audeer.list_file_names(repo, recursive=True) == expected_paths

        # Check entries of dependency table
        table_file = self.table_file(db, storage_format)
        assert deps.tables == [table_file]
        assert deps.media == [file]
        assert deps.checksum(table_file) == expected_table_checksum(
            audeer.path(build_dir, table_file)
        )
        if storage_format == "csv":
            assert deps.archive(table_file) == table
        elif storage_format == "parquet":
            assert deps.archive(table_file) == ""

    def test_database_load(
        self,
        build_dir: str,
        db: audformat.Database,
        repository: audb.Repository,
    ):
        r"""Test correct files in cache after loading database.

        Args:
            build_dir: build dir fixture
            db: database fixture
            repository: repository fixture

        """
        # Publish database
        db.save(build_dir, storage_format=storage_format)
        version = "1.0.0"
        deps = audb.publish(build_dir, version, repository)

        # Load database to cache
        db = audb.load(db.name, version=version, verbose=False, full_path=False)

        assert os.path.exists(audeer.path(db.root, self.media_file(db)))
        assert os.path.exists(audeer.path(db.root, self.table_file(db, storage_format)))
        assert not os.path.exists(
            audeer.path(db.root, self.other_table_file(db, storage_format))
        )

    def test_updated_database_save(self, build_dir, db, storage_format):
        r"""Test correct files are stored to build dir after database update.

        Args:
            build_dir: build dir fixture
            db: database fixture
            storage_format: storage format of database tables

        """
        # Publish first version
        db.save(build_dir, storage_format=storage_format)
        previous_version = "1.0.0"
        deps = audb.publish(build_dir, previous_version, repository)

        # Clear build dir to force audb.load_to() to load from backend
        audeer.rmdir(build_dir)
        # Load previous version of database
        db = audb.load_to(build_dir, db.name, version=previous_version, verbose=False)

        # Update database by adding a column
        db[table]["object"] = audformat.Column()
        db[table]["object"].set(["!!!"])
        db.save(build_dir, storage_format=storage_format)

        assert os.path.exists(
            audeer.path(build_dir, self.table_file(db, storage_format))
        )
        assert not os.path.exists(
            audeer.path(build_dir, self.other_table_file(db, storage_format))
        )
        assert os.path.exists(self.media_file(db))


    def test_updated_database_publish(
        self,
        build_dir: str,
        db: audformat.Database,
        repository: audb.Repository,
        storage_format: str,
    ):
        r"""Tests files in repository and dependency table after database update.

        Tables are stored differently in the repository
        for different storage formats,
        and their dependency table entry looks different as well.

        Args:
            build_dir: build dir fixture
            db: database fixture
            repository: repository fixture
            storage_format: database table storage format

        """
        # Publish first version
        db.save(build_dir, storage_format=storage_format)
        previous_version = "1.0.0"
        deps = audb.publish(build_dir, previous_version, repository)

        # Update database by adding a column
        db[table]["object"] = audformat.Column()
        db[table]["object"].set(["!!!"])
        db.save(build_dir, storage_format=storage_format)

        # Publish second version
        db.save(build_dir, storage_format=storage_format)
        version = "1.1.0"
        deps = audb.publish(
            build_dir,
            version,
            repository,
            previous_version=previous_version,
        )

        # Check files are published to repository
        repo = audeer.path(repository.host, repository.name)
        file = self.media_file(db)
        table = list(db)[0]
        dependency_file = "db.parquet"
        header_file = "db.yaml"
        media_file = f"{deps.archive(file)}.zip"
        if storage_format == "csv":
            meta_file = f"{table}.zip"
        elif storage_format == "parquet":
            meta_file = f"{table}.parquet"
        expected_paths = [
            audeer.path(repo, db.name, "1.0.0", dependency_file),
            audeer.path(repo, db.name, "1.0.0", header_file),
            audeer.path(repo, db.name, "1.1.0", dependency_file),
            audeer.path(repo, db.name, "1.1.0", header_file),
            audeer.path(repo, db.name, "media", "1.0.0", media_file),
            audeer.path(repo, db.name, "meta", "1.0.0", meta_file),
            audeer.path(repo, db.name, "meta", "1.1.0", meta_file),
        ]
        assert audeer.list_file_names(repo, recursive=True) == expected_paths

        # Check entries of dependency table
        table_file = self.table_file(db, storage_format)
        assert deps.tables == [table_file]
        assert deps.media == [file]
        assert deps.checksum(table_file) == expected_table_checksum(
            audeer.path(build_dir, table_file)
        )
        if storage_format == "csv":
            assert deps.archive(table_file) == table
        elif storage_format == "parquet":
            assert deps.archive(table_file) == ""

    def test_updated_database_load(
        self,
        build_dir: str,
        db: audformat.Database,
        repository: audb.Repository,
    ):
        r"""Test correct files in cache after loading updated database.

        Args:
            build_dir: build dir fixture
            db: database fixture
            repository: repository fixture

        """
        # Publish first version
        db.save(build_dir, storage_format=storage_format)
        previous_version = "1.0.0"
        deps = audb.publish(build_dir, previous_version, repository)

        # Update database by adding a column
        db[table]["object"] = audformat.Column()
        db[table]["object"].set(["!!!"])
        db.save(build_dir, storage_format=storage_format)

        # Publish second version
        db.save(build_dir, storage_format=storage_format)
        version = "1.1.0"
        deps = audb.publish(
            build_dir,
            version,
            repository,
            previous_version=previous_version,
        )
        # Publish database
        db.save(build_dir, storage_format=storage_format)
        version = "1.0.0"
        deps = audb.publish(build_dir, version, repository)

        # Load database to cache
        db = audb.load(db.name, version=version, verbose=False, full_path=False)

        assert os.path.exists(audeer.path(db.root, self.media_file(db)))
        assert os.path.exists(audeer.path(db.root, self.table_file(db, storage_format)))
        assert not os.path.exists(
            audeer.path(db.root, self.other_table_file(db, storage_format))
        )
        assert "object" in db["files"].df.columns
```

Reply:
It is true that still lots of boilerplate setup code is repeated.  But even now I find this solution helps me in the understanding as one can read each of the test cases in isolation

But this definitively is a good point. I can see two blocks of code constributing to code reduplication.

```python
       # 1.
        # Check files are published to repository
        repo = audeer.path(repository.host, repository.name)
        file = self.media_file(db)
        table = list(db)[0]

       # 2
        db.save(build_dir, storage_format=storage_format)
        previous_version = "1.0.0"
        deps = audb.publish(build_dir, previous_version, repository)
```

How much would it help to put these into separate "private" methods of the test class?

Or would it be possible to have a mechanism that encapsulates the publication of the first and second version of the database into a method or local fixture? 

But altogether I concede that this is a test that is hard to write, with lots of setup to get to the point where you can make assertions.

Reply:
> But even now I find this solution helps me in the understanding as one can read each of the test cases in isolation

OK, I see the point.

I will try to improve the code by adding more methods, and will push it then.

Reply:
I found a way to further reduce repeating code by adding some common functions at the top of `test_publish_table.py`:

* `assert_db_published_to_repo()`
* `assert_db_saved_to_dir()`
* `assert_dependency_table()`
* `publish_db()`
* `update_db()`

The code is now integrated into this branch, and you can take another look at it.

Reply:
Wow, this is nice.  I have never seen it like that, so probably it is quite uncommon to put so much logic into assertion functions. But in terms of readability the new structure is great.  I find this lots easier to read that the imperative coding style that was there intitially. Together with isloating and documenting the logic to publish and update this helps a lot in terms of readability. 

This can be resolved here and I will proceed with the general review.

Reply:
Thanks also for insisting on improvements here, as I also think it is much easier to read now.